### DISCUSSION TITLE
how  to configure  superset  with   keycloak

### DISCUSSION THREAD

**Discussion by ouadhi** [apache/superset#13915] [Q&A / Help] [answered]
I am trying to add keycloak  authentication  any  recommendation

**Comment by srinify** [comment]
Hi @ouadhi there aren't any templates off-hand for this in the Superset documentation (superset.apache.org/docs) but this StackOverflow answer might be helpful:

https://stackoverflow.com/a/47787279/963203

If you end up getting it working, we'd love to see you contribute the recipe to the documentation!

**Comment by ouadhi** [comment, marked as answer]
hi @srinify  , it works perfectly now , 
first  ,  i add flask-oidc==1.3.0   in  requirement.test  
in docker/pythonpath_dev create file  kyecloack_securtiy_manager.py  and  add  this code  
```python
rom flask_appbuilder.security.manager import AUTH_OID
from superset.security import SupersetSecurityManager
from flask_oidc import OpenIDConnect
from flask_appbuilder.security.views import AuthOIDView
from flask_login import login_user
from urllib.parse import quote
from flask_appbuilder.views import ModelView, SimpleFormView, expose
import logging

class OIDCSecurityManager(SupersetSecurityManager):

    def __init__(self, appbuilder):
        super(OIDCSecurityManager, self).__init__(appbuilder)
        if self.auth_type == AUTH_OID:
            self.oid = OpenIDConnect(self.appbuilder.get_app)
        self.authoidview = AuthOIDCView

class AuthOIDCView(AuthOIDView):

    @expose('/login/', methods=['GET', 'POST'])
    def login(self, flag=True):
        sm = self.appbuilder.sm
        oidc = sm.oid

        @self.appbuilder.sm.oid.require_login
        def handle_login():
            user = sm.auth_user_oid(oidc.user_getfield('email'))

            if user is None:
                info = oidc.user_getinfo(['preferred_username', 'given_name', 'family_name', 'email'])
                user = sm.add_user(info.get('preferred_username'), info.get('given_name'), info.get('family_name'),
                                   info.get('email'), sm.find_role('Gamma'))

            login_user(user, remember=False)
            return redirect(self.appbuilder.get_url_for_index)

        return handle_login()

    @expose('/logout/', methods=['GET', 'POST'])
    def logout(self):
        oidc = self.appbuilder.sm.oid

        oidc.logout()
        super(AuthOIDCView, self).logout()
        redirect_url = request.url_root.strip('/') + self.appbuilder.get_url_for_login

        return redirect(
            oidc.client_secrets.get('issuer') + '/protocol/openid-connect/logout?redirect_uri=' + quote(redirect_url))
```
create  another json file  "client_secret.json" contains keycloack configuration 

```json
{
    "web": {
        "issuer": "http://keyclaokdomain/auth/realms/<realmName>",
        "auth_uri": "http://keyclaokdomain/auth/realms/<realmName>/protocol/openid-connect/auth",
        "client_id": "<ClientID>",
        "client_secret": "<Client Secret>",
        "redirect_uris": [
            "http://domaineApp/*"
        ],
        "userinfo_uri": "http://keyclaokdomain/auth/realms/<realmName>/protocol/openid-connect/userinfo",
        "token_uri": "http://keyclaokdomain/auth/realms/<realmName>/protocol/openid-connect/token",
        "token_introspection_uri": "http://keyclaokdomain/auth/realms/<realmName>/protocol/openid-connect/token/introspect"
    }
}

```
finally in superset_config.py  add  lines  
```python
from kyecloack_securtiy_manager  import  OIDCSecurityManager
from flask_appbuilder.security.manager import AUTH_OID, AUTH_REMOTE_USER, AUTH_DB, AUTH_LDAP, AUTH_OAUTH
import os
'''
---------------------------KEYCLOACK ----------------------------
'''
curr  =  os.path.abspath(os.getcwd())
AUTH_TYPE = AUTH_OID
SECRET_KEY: 'SomethingNotEntirelySecret'
OIDC_CLIENT_SECRETS =  curr + '/pythonpath/client_secret.json'
OIDC_ID_TOKEN_COOKIE_SECURE = False
OIDC_REQUIRE_VERIFIED_EMAIL = False
OIDC_OPENID_REALM: 'realm1'
OIDC_INTROSPECTION_AUTH_METHOD: 'client_secret_post'
CUSTOM_SECURITY_MANAGER = OIDCSecurityManager
AUTH_USER_REGISTRATION = True
AUTH_USER_REGISTRATION_ROLE = 'Gamma'
'''
--------------------------------------------------------------
'''
```

**Reply by voriol** [reply]
hi @ouadhi, thanks for the code, it works fine, just one thing, I had to add an import for the solution to work 100%:

`from flask import redirect, request`

(keycloak_securtiy_manager.py)

**Reply by cemremengu** [reply]
UPDATE: Ok so `sm.add_user` returns False when there is an error inserting the user hence the error

I get error with same config

```
        Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/etc/superset/keycloak_security_manager.py", line 38, in login
    return handle_login()
  File "/usr/local/lib/python3.7/site-packages/flask_oidc/__init__.py", line 484, in decorated
    return view_func(*args, **kwargs)
  File "/etc/superset/keycloak_security_manager.py", line 35, in handle_login
    login_user(user, remember=False)
  File "/usr/local/lib/python3.7/site-packages/flask_login/utils.py", line 158, in login_user
    if not force and not user.is_active:
AttributeError: 'bool' object has no attribute 'is_active'
```

Any ideas?

**Reply by ouadhi** [reply]
Hi , @cemremengu  
in login_user function set `login_user(user, remember=False ,force=True)`

**Reply by cemremengu** [reply]
I think force helps when user is inactive. In my case same username (admin) existed with different email in superset so it tried to recreate it but there was a PK error. Trying a different user solved the problem.

Thank you for the help!

**Reply by ciolo** [reply]
@ouadhi It works like a charm. But how do you log in with a user with the Admin role so you can create datasets, dashboards, charts etc? I would like to have a user with administrator privileges and other guest users who can only see some graphs or dashboards, as it works in the case of default login via database.
Thank u.

**Reply by wiktor2200** [reply]
The easiest method to do it is create Admin user with CLI while provisioning superset for the first time and define your custom role and set is a default in keycloak script as above.
The other method is to run config for Admin for the first time, create everything and then change default role to yours.

**Reply by cemremengu** [reply]
Note that due to some breaking changes you need to set `itsdangerous==2.0.1` for `flask_oidc`. I also had to install `flask_openid `

**Reply by aidatabases** [reply]
### In superset_config.py
### This works: OIDC_CLIENT_SECRETS = curr + '/docker/pythonpath_dev/client_secret.json'
### instead of: OIDC_CLIENT_SECRETS =  curr + '/pythonpath/client_secret.json'
--
Btw thanks, it really helped my use case

**Reply by rahul149386** [reply]
@ouadhi, @aidatabases @cemremengu  I followed your comments, made the changes accordingly. 
 but when I visit the http://:8088/login it is redirecting me to keycloack for openId auth. When we give credentials in keycloak page. noticing keycloak able send to superset oidc_callback with GET/oidc_callback?state=XXXXX. Keycloak and superset installed on same DNS.
![image](https://user-images.githubusercontent.com/56782848/197343334-481f82b4-6e31-46b4-913d-f8e789971b23.png)

But I'm getting below error. Anyone kindly help me to resolve this 
<img width="926" alt="image" src="https://user-images.githubusercontent.com/56782848/197342687-a61aba39-7f6b-49cf-8816-82af6c7aa1fe.png">
<img width="762" alt="image" src="https://user-images.githubusercontent.com/56782848/197342733-6cfcf1b4-b15c-426d-b6e4-9bec77dcae2f.png">

**Reply by saatvikrk** [reply]
After the code change I am able to see the keycloak login screen
![image](https://github.com/apache/superset/assets/133918920/99134a79-f0cd-4068-9e64-286915d86e1f)

After providing the credentials I am seeing the below error

![image](https://github.com/apache/superset/assets/133918920/0d37b4da-fe13-42b6-8cf6-c82a292e3d9d)

Any help would be greatly appreciated.

**Reply by saatvikrk** [reply]
> @ouadhi, @aidatabases @cemremengu I followed your comments, made the changes accordingly. but when I visit the http://:8088/login it is redirecting me to keycloack for openId auth. When we give credentials in keycloak page. noticing keycloak able send to superset oidc_callback with GET/oidc_callback?state=XXXXX. Keycloak and superset installed on same DNS. ![image](https://user-images.githubusercontent.com/56782848/197343334-481f82b4-6e31-46b4-913d-f8e789971b23.png)
> 
> But I'm getting below error. Anyone kindly help me to resolve this <img alt="image" width="926" src="https://user-images.githubusercontent.com/56782848/197342687-a61aba39-7f6b-49cf-8816-82af6c7aa1fe.png"> <img alt="image" width="762" src="https://user-images.githubusercontent.com/56782848/197342733-6cfcf1b4-b15c-426d-b6e4-9bec77dcae2f.png">

@Rahul, are you able to fix the connectivity issue you mentioned acove ?

**Reply by saatvikrk** [reply]
> After the code change I am able to see the keycloak login screen ![image](https://user-images.githubusercontent.com/133918920/239076785-99134a79-f0cd-4068-9e64-286915d86e1f.png)
> 
> After providing the credentials I am seeing the below error
> 
> ![image](https://user-images.githubusercontent.com/133918920/239076952-0d37b4da-fe13-42b6-8cf6-c82a292e3d9d.png)
> 
> Any help would be greatly appreciated.

My keycloak is running outside of docker and superset is running inside docker container , so  when keycloak redirect to superset after authentication I am still calling the localhost rather than calling container ip running in the docker.

**Comment by Synarcs** [comment]
@ouadhi I followed the same but when I visit the http://<superset-domain>:8088/login it is not redirecting me to keycloack for openId auth rather its appending URL like http://<superset-domain>:8088/login/<keycloak_domain>:8080/<more keyclaok meta opeid meta conf>. are there any specific config needed n keyclaok  because the oidc  callback url is not redirecting to superset home page.
I am using jboss/keycloak latest docker image
Docker based superset version 1.1.0
Please can someone help regarding this issue

**Reply by ouadhi** [reply]
hi @vedangparasnis  maybe is less configuration in your keycloak and i think it is necessary when you use Docker add host configuration in docker run

**Reply by Synarcs** [reply]
@ouadhi i tried used a barebone keycloak setup using standalone.sh -b 0.0.0.0 and my keycloak started on host ip of instance. My Superset is running on a different instance with default docker-compose-non-dev.yml in superset.
But When i hit http://<superset Instance Ip>:8088/
It did append the same thing like 
http://10.234.55.136:8088/10.234.57.13:8080/auth/realms/realm1/protocol/openid-connect/auth?client_id=sys&redirect_uri=http%3A%2F%2F10.234.55.136%3A8088%2Foidc_callback&scope=openid+email&access_type=offline&response_type=code&state=eyJjc3JmX3Rva2VuIjogInhGbU0zSzJwWkJtQ0p5SXRzZXFhamxVejlnNTFjZk5xIiwgImRlc3RpbmF0aW9uIjogImV5SmhiR2NpT2lKSVV6VXhNaUo5LkltaDBkSEE2THk4eE1DNHlNelF1TlRVdU1UTTJPamd3T0RndmJHOW5hVzR2SWcuNnAwdGlUWS1mVW1ZN281Z1F3WWxXdm1hVVFZQkw3Sl9HVjFCdGNnbm5IYnM4THI1bWk1SjJ0SmF5ZWV3dDVuQTJNY1pCcHY5UEUxVG5RTG9fLXg5NVEifQ%3D%3D

10.234.55.136 --> Superset Instance Ip 
10.234.57.13   --> Keycloak IP

**Reply by rahul149386** [reply]
@vedangparasnis did you resolve your issue.If yes, let me know

**Comment by wiktor2200** [comment]
Hi! Is there anyone here who have tried pass: `"client_secret": "<Client Secret>",` using Vault by HashiCorp?
I'm facing some problems, I've added annotations for pods and vault should be accessible. Nevertheless I'm getting unauthorized error:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/app/superset_home/.local/lib/python3.8/site-packages/flask_oidc/__init__.py", line 650, in _oidc_callback
    credentials = flow.step2_exchange(code)
  File "/app/superset_home/.local/lib/python3.8/site-packages/oauth2client/_helpers.py", line 133, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/app/superset_home/.local/lib/python3.8/site-packages/oauth2client/client.py", line 2089, in step2_exchange
    raise FlowExchangeError(error_msg)
oauth2client.client.FlowExchangeError: unauthorized_clientInvalid client secret
```
The whole flow of keycloak works well, I've tried to add `client_secret` with plain text and it works.

**Comment by nathanielread** [comment]
I may have done something wrong, but to get around an error:

`ModuleNotFoundError: No module named 'flask_oidc'`

I added this line:

`pip install flask-oidc==1.3.0`

in the docker-bootstrap.sh file about here:
https://github.com/apache/superset/blob/master/docker/docker-bootstrap.sh#L36

To check it is running, add an echo before it like:

`echo "Installing flask_oidc"`

Hope that helps someone else. 

Or if someone can tell me why I don't need to do that and should instead do another thing, please let me know.

Otherwise, I'm able to login with keycloak and it adds my user from the login.

**Reply by wiktor2200** [reply]
Hi! In post: https://github.com/apache/superset/discussions/13915#discussioncomment-558709
It was mentioned in second line:
> first , i add flask-oidc==1.3.0 in requirement.test

flask-oidc is needed to OIDC in general and as it is module you have to attach it only when needed :)

When using helm-chart you can place it there: https://github.com/apache/superset/blob/master/helm/superset/values.yaml#L38

**Reply by nathanielread** [reply]
Thanks, @wiktor2200 .

Where's the correct place when using docker?

**Reply by wiktor2200** [reply]
I guess in `./docker/requirements-local.txt` file.
Here's the manual: https://github.com/apache/superset/tree/master/docker#local-packages

Then the file which you will create is defined here: [LINK1](https://github.com/apache/superset/blob/9671384efebb2cf77fcd9bedae63192c0cf6570d/docker/docker-bootstrap.sh#L21) and used in this if clause: [LINK2](https://github.com/apache/superset/blob/9671384efebb2cf77fcd9bedae63192c0cf6570d/docker/docker-bootstrap.sh#L32)

**Comment by marvincorreia** [comment]
I managed to solve the integration and Its working like a charm, check the code in github [superset-keycloak](https://github.com/marvincorreia/superset-keycloak)

**Reply by mgorsk1** [reply]
is this compatible with Superset 2?

**Reply by wiktor2200** [reply]
Hi @mgorsk1! Keycloak configuration works in Superset 2.0.0.
Some dependencies has changed (`itsdangerous` deprecated some function so it has to be installed in version <2.1.0) as mentioned in this comment: https://github.com/apache/superset/discussions/13915#discussioncomment-2542250.

**Reply by marvincorreia** [reply]
@mgorsk1 Yes It is!

**Reply by mgorsk1** [reply]
Thanks for replies @wiktor2200 @marvincorreia . I actually figured out how to use keycloak oidc without `fab-oidc` which I think is the better way. Assuming you use `flask-appbuilder>=4.1.3` you can simply:

```
  from flask_appbuilder.security.manager import AUTH_OAUTH
  from superset.security import SupersetSecurityManager

  # Set the authentication type to OAuth
  AUTH_TYPE = AUTH_OAUTH

  KEYCLOAK_URL = os.getenv('KEYCLOAK_URL')
  KEYCLOAK_REALM = os.getenv('KEYCLOAK_REALM')

  CLIENT_ID = os.getenv('KEYCLOAK_CLIENT_ID')
  CLIENT_SECRET = os.getenv('KEYCLOAK_CLIENT_SECRET')

  OAUTH_PROVIDERS = [
      {   'name':'keycloak',
          'token_key':'access_token',
          'icon':'fa-icon',
          'remote_app': {
              'client_id': CLIENT_ID,
              'client_secret': CLIENT_SECRET,
              'client_kwargs': {
                  'scope': 'openid email profile'
              },
              'access_token_method':'POST',
              'api_base_url': f'{KEYCLOAK_URL}/realms/{KEYCLOAK_REALM}/protocol/',
              'access_token_url': f'{KEYCLOAK_URL}/realms/{KEYCLOAK_REALM}/protocol/openid-connect/token',
              'authorize_url': f'{KEYCLOAK_URL}/realms/{KEYCLOAK_REALM}/protocol/openid-connect/auth',
              'jwks_uri': f'{KEYCLOAK_URL}/realms/{KEYCLOAK_REALM}/protocol/openid-connect/certs',
              'request_token_url': None,
          }
      }
  ]

  # Will allow user self registration, allowing to create Flask users from Authorized User
  AUTH_USER_REGISTRATION = True
```

keycloak provider was introducer here: https://github.com/dpgaspar/Flask-AppBuilder/pull/1832

The only caveat is that it uses `preferred_username` instead of `sub` claim for user id (which was default for `fab-oidc`) and sadly it's not configurable yet but quite an easy thing to contribute if needed.

Note that urls in `OAUTH_PROVIDERS` do not contain `/auth` path as we are on quite recent version of keycloak which has removed it.

**Reply by exherb** [reply]
how to config SupersetSecurityManager?

**Reply by mgorsk1** [reply]
what exactly you want to configure? exemplary oidc flask appbuilder SecurityManager (which is what SupersetSecurityManager is built on) configuration can be found here: https://github.com/dpgaspar/Flask-AppBuilder/blob/master/examples/oauth/config.py

**Reply by ancogamer** [reply]
> what exactly you want to configure? exemplary oidc flask appbuilder SecurityManager (which is what SupersetSecurityManager is built on) configuration can be found here: https://github.com/dpgaspar/Flask-AppBuilder/blob/master/examples/oauth/config.py

Hi can you show a example about where this code should go ?

**Reply by Hokwang** [reply]
@mgorsk1 do you use keycloak group mapping to superset role ? if then, please let me know.

**Reply by mgorsk1** [reply]
not really @Hokwang, we have default superset role assigned on account creation (first login) and then every user has same role, access to data is achieved through user impersonation.

**Reply by mgorsk1** [reply]
@ancogamer this code should go to superset_config.py module that is available under PYTHONPATH when running superset

**Reply by ancogamer** [reply]
> @ancogamer this code should go to superset_config.py module that is available under PYTHONPATH when running superset

Thanks :)
